File: exercises/day19-North-Pole-Companies/java/src/test/java/com/northpole/InvoicePrinterTest.java
package com.northpole;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;

class InvoicePrinterTest {

    @Test
    void exampleInvoice() throws IOException {
        Map<String, ElfCompany> elfCompanies = loadElfCompanies();
        Invoice invoice = loadInvoice();
        InvoicePrinter printer = new InvoicePrinter();
        
        String result = printer.print(invoice, elfCompanies);
        
        Approvals.verify(result);
    }

    // TODO: Add exampleInvoiceWithTaxes() test here

    private Map<String, ElfCompany> loadElfCompanies() throws IOException {
        String json = Files.readString(Paths.get("src/test/resources/elfCompanies.json"));
        Gson gson = new Gson();
        Map<String, Map<String, String>> rawData = gson.fromJson(json, 
            new TypeToken<Map<String, Map<String, String>>>(){}.getType());
        
        Map<String, ElfCompany> companies = new HashMap<>();
        for (Map.Entry<String, Map<String, String>> entry : rawData.entrySet()) {
            Map<String, String> data = entry.getValue();
            companies.put(entry.getKey(), 
                new ElfCompany(data.get("name"), data.get("type"), data.get("region")));
        }
        return companies;
    }

    private Invoice loadInvoice() throws IOException {
        String json = Files.readString(Paths.get("src/test/resources/order.json"));
        Gson gson = new Gson();
        Map<String, Object> orderData = gson.fromJson(json, 
            new TypeToken<Map<String, Object>>(){}.getType());
        
        String customer = (String) orderData.get("customer");
        List<Map<String, Object>> deliveriesData = 
            (List<Map<String, Object>>) orderData.get("deliveries");
        
        List<Delivery> deliveries = new ArrayList<>();
        for (Map<String, Object> deliveryData : deliveriesData) {
            deliveries.add(new Delivery(
                (String) deliveryData.get("companyID"),
                ((Double) deliveryData.get("packages")).intValue()
            ));
        }
        return new Invoice(customer, deliveries);
    }
}

File: exercises/day19-North-Pole-Companies/kotlin/src/test/kotlin/com/northpole/InvoicePrinterTest.kt
package com.northpole

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import org.approvaltests.Approvals
import org.junit.jupiter.api.Test
import java.nio.file.Files
import java.nio.file.Paths

class InvoicePrinterTest {

    @Test
    fun exampleInvoice() {
        val elfCompanies = loadElfCompanies()
        val invoice = loadInvoice()
        val printer = InvoicePrinter()

        val result = printer.print(invoice, elfCompanies)

        Approvals.verify(result)
    }

    // TODO: Add exampleInvoiceWithTaxes test here

    private fun loadElfCompanies(): Map<String, ElfCompany> {
        val json = Files.readString(Paths.get("src/test/resources/elfCompanies.json"))
        val gson = Gson()
        val type = object : TypeToken<Map<String, Map<String, String>>>() {}.type
        val rawData: Map<String, Map<String, String>> = gson.fromJson(json, type)

        return rawData.mapValues { (_, data) ->
            ElfCompany(
                name = data["name"]!!,
                type = data["type"]!!,
                region = data["region"]!!
            )
        }
    }

    private fun loadInvoice(): Invoice {
        val json = Files.readString(Paths.get("src/test/resources/order.json"))
        val gson = Gson()
        val type = object : TypeToken<Map<String, Any>>() {}.type
        val orderData: Map<String, Any> = gson.fromJson(json, type)

        val customer = orderData["customer"] as String
        val deliveriesData = orderData["deliveries"] as List<Map<String, Any>>

        val deliveries = deliveriesData.map { deliveryData ->
            Delivery(
                companyID = deliveryData["companyID"] as String,
                packages = (deliveryData["packages"] as Double).toInt()
            )
        }

        return Invoice(customer, deliveries)
    }
}
